ocaml/xapi/xapi.ml

```diff
@@ -941,22 +941,34 @@ let server_init() =
                                    (fun () -> Debug_populate.do_populate ~vms:1000 ~vdis_per_vm:3 ~networks:10 ~srs:10 ~tasks:1000) ]
         );
 
+        let maybe_resync_cluster_host () =
+          let host = Helpers.get_localhost ~__context in
+          begin match Xapi_clustering.find_cluster_host ~__context ~host with
+          | Some self -> ignore
+            Xapi_mgmt_iface.(wait_for_clustering_ip ~__context ~self);
+            Xapi_cluster_host.resync_host ~__context ~host
+          | None -> () end
+        in
+
         let wait_management_interface () =
           let management_if = Xapi_inventory.lookup Xapi_inventory._management_interface in
-          if management_if <> "" then (
+          if management_if <> "" then begin
             debug "Waiting forever for the management interface to gain an IP address";
             let ip = wait_for_management_ip_address ~__context in
             debug "Management interface got IP address: %s; attempting to re-plug any unplugged PBDs" ip;
+            (* CA-290237: If any clusters exist, we want to ensure we have an IP address before enabling cluster hosts *)
+            log_and_ignore_exn (fun () -> maybe_resync_cluster_host ());
+            (* Now we can initialise any clustered storage *)
             Helpers.call_api_functions ~__context (fun rpc session_id ->
                 Create_storage.plug_unplugged_pbds __context)
-          )
+          end
         in
 
         Startup.run ~__context [
           "fetching database backup", [ Startup.OnlySlave; Startup.NoExnRaising ],
           (fun () -> Pool_db_backup.fetch_database_backup ~master_address:(Pool_role.get_master_address())
               ~pool_secret:!Xapi_globs.pool_secret ~force:None);
-          "wait management interface to come up", [ Startup.NoExnRaising ], wait_management_interface;
+          "wait management interface to come up, resync any cluster_hosts created", [ Startup.NoExnRaising ], wait_management_interface;
```

Comment:
Since this is a blocking action, if something goes wrong and the networking interface of clustering doesn't come up, it would be nice if we didn't block the whole of XAPI.
So perhaps introduce a second entry here below 'wait for management interface', that says "wait for clustering IP if any, and plug unplugged PBDs" that calls your `maybe_wait_for_clustering_ip` function above.